File: dotnet-desktop-guide/framework/winforms/advanced/snippets/how-to-create-a-private-font-collection/cs/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Text;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace fonts
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        // <DrawFont>
        // Helper function to print text in a font and style
        private float DrawFont(Graphics graphicsObj,
                               FontFamily family,
                               FontStyle style,
                               SolidBrush colorBrush,
                               PointF location,
                               string styleName)
        {
            // The string to print, which contains the family name and style
            string familyNameAndStyle = $"{family.Name} {styleName}";

            // Create the font object
            using (Font fontObject = new Font(family.Name, 16, style, GraphicsUnit.Pixel))
            {
                // Draw the string
                graphicsObj.DrawString(familyNameAndStyle, fontObject, colorBrush, location);

                // Return the height of the font
                return fontObject.Height;
            }
        }

        // The OnPaint method of a form, which provides the graphics object
        protected override void OnPaint(PaintEventArgs e)
        {
            PointF location = new PointF(10, 0);
            SolidBrush solidBrush = new SolidBrush(Color.Black);

            FontFamily[] fontFamilies;
            PrivateFontCollection privateFontCollection = new PrivateFontCollection(); // Dispose later

            // Add three font files to the private collection.
            privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\\Fonts\\Arial.ttf"));
            privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\\Fonts\\CourBI.ttf"));
            privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\\Fonts\\TimesBD.ttf"));

            // Get the array of FontFamily objects.
            fontFamilies = privateFontCollection.Families;

            // Process each font in the collection
            for (int i = 0; i < fontFamilies.Length; i++)
            {
                // Draw the font in every style

                // Regular
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Regular))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Regular, solidBrush, location, "Regular");

                // Bold
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Bold))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Bold, solidBrush, location, "Bold");

                // Italic
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Italic))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Italic, solidBrush, location, "Italic");

                // Bold and Italic
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Bold) &&
                    fontFamilies[i].IsStyleAvailable(FontStyle.Italic))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Bold | FontStyle.Italic, solidBrush, location, "BoldItalic");

                // Underline
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Underline))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Underline, solidBrush, location, "Underline");

                // Strikeout
                if (fontFamilies[i].IsStyleAvailable(FontStyle.Strikeout))
                    location.Y += DrawFont(e.Graphics, fontFamilies[i], FontStyle.Strikeout, solidBrush, location, "Strikeout");

                // Extra space between font families
                location.Y += 10;
            }

            privateFontCollection.Dispose();
        }
        // </DrawFont>
    }
}

File: dotnet-desktop-guide/framework/winforms/advanced/snippets/how-to-create-a-private-font-collection/vb/Form1.vb
Imports System.Drawing.Text

Public Class Form1

    '<DrawFont>
    ' Helper function to print text in a font and style
    Private Function DrawFont(graphicsObj As Graphics,
                              family As FontFamily,
                              style As FontStyle,
                              colorBrush As SolidBrush,
                              location As PointF,
                              styleName As String) As Single

        ' The string to print, which contains the family name and style
        Dim familyNameAndStyle As String = $"{family.Name} {styleName}"

        ' Create the font object
        Using fontObject As New Font(family.Name, 16, style, GraphicsUnit.Pixel)

            ' Draw the string
            graphicsObj.DrawString(familyNameAndStyle, fontObject, colorBrush, location)

            ' Return the height of the font
            Return fontObject.Height

        End Using

    End Function

    ' The OnPaint method of a form, which provides the graphics object
    Protected Overrides Sub OnPaint(e As PaintEventArgs)

        Dim location As New PointF(10, 0)
        Dim solidBrush As New SolidBrush(Color.Black)

        Dim fontFamilies() As FontFamily
        Dim privateFontCollection As New PrivateFontCollection() ' Dispose later

        ' Add three font files to the private collection.
        privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\Fonts\Arial.ttf"))
        privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\Fonts\CourBI.ttf"))
        privateFontCollection.AddFontFile(System.Environment.ExpandEnvironmentVariables("%systemroot%\Fonts\TimesBD.ttf"))

        ' Get the array of FontFamily objects.
        fontFamilies = privateFontCollection.Families

        ' Process each font in the collection
        For i = 0 To fontFamilies.Length - 1

            ' Draw the font in every style

            ' Regular
            If fontFamilies(i).IsStyleAvailable(FontStyle.Regular) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Regular, solidBrush, location, "Regular")
            End If

            ' Bold
            If fontFamilies(i).IsStyleAvailable(FontStyle.Bold) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Bold, solidBrush, location, "Bold")
            End If

            ' Italic
            If fontFamilies(i).IsStyleAvailable(FontStyle.Italic) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Italic, solidBrush, location, "Italic")
            End If

            ' Bold and Italic
            If fontFamilies(i).IsStyleAvailable(FontStyle.Italic) And
               fontFamilies(i).IsStyleAvailable(FontStyle.Italic) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Bold Or FontStyle.Italic, solidBrush, location, "BoldItalic")
            End If

            ' Underline
            If fontFamilies(i).IsStyleAvailable(FontStyle.Underline) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Underline, solidBrush, location, "Underline")
            End If

            ' Strikeout
            If fontFamilies(i).IsStyleAvailable(FontStyle.Strikeout) Then
                location.Y += DrawFont(e.Graphics, fontFamilies(i), FontStyle.Strikeout, solidBrush, location, "Strikeout")
            End If

            ' Extra space between font families
            location.Y += 10

        Next

        privateFontCollection.Dispose()

    End Sub
    '</DrawFont>

End Class
